#Hw3

##2015310908
##임성규

In [1]:
from google.colab import drive 
drive.mount('/content/gdrive/')

Mounted at /content/gdrive/


# dataset

원본 데이터를 가져온뒤 item들인 winning number만 따로 추출하여 알고리즘에 사용하기 적합한 형태로 가공하였습니다.

In [2]:
import pandas as pd
df = pd.read_excel("/content/gdrive/MyDrive/hw3/lotto645-1.xlsx")

df

회차별 추첨결과 Unnamed: 1  Unnamed: 2  ... Unnamed: 17 Unnamed: 18 Unnamed: 19
0         년도         회차         추첨일  ...         NaN         NaN         NaN
1        NaN        NaN         NaN  ...         5.0         6.0         보너스
2       2020        939  2020.11.28  ...        42.0        45.0           6
3        NaN        938  2020.11.21  ...        31.0        36.0           9
4        NaN        937  2020.11.14  ...        29.0        40.0          26
..       ...        ...         ...  ...         ...         ...         ...
936      NaN          5  2003.01.04  ...        41.0        42.0           3
937     2002          4  2002.12.28  ...        40.0        42.0           2
938      NaN          3  2002.12.21  ...        27.0        31.0          30
939      NaN          2  2002.12.14  ...        32.0        42.0           2
940      NaN          1  2002.12.07  ...        37.0        40.0          16

[941 rows x 20 columns]

In [3]:
df_winningnum = df.iloc[2:, 13:20]

print(df_winningnum)

df_transactions = df.iloc[2:, 1:2]

df_transactions.head()

    Unnamed: 13  Unnamed: 14  Unnamed: 15  ...  Unnamed: 17  Unnamed: 18  Unnamed: 19
2             4         11.0         28.0  ...         42.0         45.0            6
3             4          8.0         10.0  ...         31.0         36.0            9
4             2         10.0         13.0  ...         29.0         40.0           26
5             7         11.0         13.0  ...         18.0         29.0           43
6             4         10.0         20.0  ...         38.0         44.0           18
..          ...          ...          ...  ...          ...          ...          ...
936          16         24.0         29.0  ...         41.0         42.0            3
937          14         27.0         30.0  ...         40.0         42.0            2
938          11         16.0         19.0  ...         27.0         31.0           30
939           9         13.0         21.0  ...         32.0         42.0            2
940          10         23.0         29.0  ...        

Unnamed: 1
2        939
3        938
4        937
5        936
6        935

제가 사용하고자하는 apriori 함수를 사용하려면 transactions를 list형으로 사용해야하여 아래코드를 통해 변형시켜주었습니다.

In [40]:
transactions = []
for i in range(0, len(df_winningnum)):
    transactions.append([str(df_winningnum.values[i,j]) for j in range(0, 7)])

print(transactions)

[['4', '11.0', '28.0', '39.0', '42.0', '45.0', '6'], ['4', '8.0', '10.0', '16.0', '31.0', '36.0', '9'], ['2', '10.0', '13.0', '22.0', '29.0', '40.0', '26'], ['7', '11.0', '13.0', '17.0', '18.0', '29.0', '43'], ['4', '10.0', '20.0', '32.0', '38.0', '44.0', '18'], ['1', '3.0', '30.0', '33.0', '36.0', '39.0', '12'], ['23', '27.0', '29.0', '31.0', '36.0', '45.0', '37'], ['1', '6.0', '15.0', '36.0', '37.0', '38.0', '5'], ['14', '15.0', '23.0', '25.0', '35.0', '43.0', '32'], ['8', '21.0', '25.0', '38.0', '39.0', '44.0', '28'], ['7', '9.0', '12.0', '15.0', '19.0', '23.0', '4'], ['3', '4.0', '10.0', '20.0', '28.0', '44.0', '30'], ['4', '15.0', '22.0', '38.0', '41.0', '43.0', '26'], ['10', '16.0', '18.0', '20.0', '25.0', '31.0', '6'], ['13', '24.0', '32.0', '34.0', '39.0', '42.0', '4'], ['3', '11.0', '34.0', '42.0', '43.0', '44.0', '13'], ['3', '17.0', '18.0', '23.0', '36.0', '41.0', '26'], ['2', '6.0', '13.0', '17.0', '27.0', '43.0', '36'], ['5', '7.0', '12.0', '22.0', '28.0', '41.0', '1'], ['

I included bonus number in calculating association rules.

#problem 1

python의 apyori 라이브러리에 있는 apriori 함수를 사용하여 One-antecedent-one-consequent rules를 구하였습니다.

In [5]:
!pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-cp36-none-any.whl size=5977 sha256=5c910fb17454938032cd994fc5376adfb736c22d0c5b244a46df158a9b41e6c0
  Stored in directory: /root/.cache/pip/wheels/5d/92/bb/474bbadbc8c0062b9eb168f69982a0443263f8ab1711a8cad0
Successfully built apyori


In [6]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from apyori import apriori

앞서 구축한 transactions와 함께 min_support, min_confidence, min_lift, max_length(itemset안에 item최대 개수)를 선언하여 그해 부합하는 rules들을 구하였습니다.
<br/>min_support = 0.005 (대략 939회의 추첨 중 5번의 추첨에서 숫자가 나왔을 경우)
<br/>min_confidence = 0.2 (20%)
<br/>min_lift = 2.7
<br/>max_length = 2 (One-antecedent-one-consequent rules)
<br/>min_length = 2

In [218]:
association_rules1 = apriori(transactions, min_support=0.005, min_confidence=0.2, min_lift=2.7, max_length=2) #  setting min_support, max item lengthin itemset
association_results1 = list(association_rules1)

association_results1 = list(filter(lambda x: len(x.items) > 1, association_results1)) # item num in itemset is bigger than 1

In [219]:
print(len(association_results1))

12


In [220]:
print(association_results1[0])

RelationRecord(items=frozenset({'1', '2.0'}), support=0.02023429179978701, ordered_statistics=[OrderedStatistic(items_base=frozenset({'2.0'}), items_add=frozenset({'1'}), confidence=1.0, lift=5.943037974683544)])





Apriori library를 사용하게 되면

결과 값이 frozenset이라는 자료형으로 저장이 됩니다.

위의 결과값 첫번째 rule을 출력해보면 frozenset의 구성을 확인할 수 있고 인덱싱을 통해

각 rule과 support, confidence lift값을 아래와 같이 출력하였습니다.

In [221]:
for item in association_results1:

    print("Rule: " + str(item[2][0][0]) + " -> " + str(item[2][0][1]))
    print("Support: " + str(item[1]))
    print("Confidence: " + str(item[2][0][2]))
    print("Lift: " + str(item[2][0][3]))
    print("----------------------------------------------")

Rule: frozenset({'2.0'}) -> frozenset({'1'})
Support: 0.02023429179978701
Confidence: 1.0
Lift: 5.943037974683544
----------------------------------------------
Rule: frozenset({'3.0'}) -> frozenset({'1'})
Support: 0.024494142705005325
Confidence: 0.6764705882352942
Lift: 4.020290394638868
----------------------------------------------
Rule: frozenset({'44'}) -> frozenset({'11.0'})
Support: 0.005324813631522897
Confidence: 0.29411764705882354
Lift: 2.907120743034056
----------------------------------------------
Rule: frozenset({'38'}) -> frozenset({'12.0'})
Support: 0.008519701810436636
Confidence: 0.33333333333333337
Lift: 3.068627450980393
----------------------------------------------
Rule: frozenset({'21'}) -> frozenset({'16.0'})
Support: 0.009584664536741214
Confidence: 0.34615384615384615
Lift: 3.218202589489718
----------------------------------------------
Rule: frozenset({'19'}) -> frozenset({'30.0'})
Support: 0.010649627263045794
Confidence: 0.3448275862068966
Lift: 2.891009

##  코드를 통해 얻은 rules들 중 상위 10개의 rules를 lift순으로 정리하면 다음과 같습니다.


In [ ]:
Rule: frozenset({'2.0'}) -> frozenset({'1'})
Support: 0.02023429179978701
Confidence: 1.0
Lift: 5.943037974683544
----------------------------------------------
Rule: frozenset({'3.0'}) -> frozenset({'1'})
Support: 0.024494142705005325
Confidence: 0.6764705882352942
Lift: 4.020290394638868
----------------------------------------------
Rule: frozenset({'45'}) -> frozenset({'8.0'})
Support: 0.005324813631522897
Confidence: 0.33333333333333337
Lift: 3.9125000000000005
----------------------------------------------
Rule: frozenset({'21'}) -> frozenset({'16.0'})
Support: 0.009584664536741214
Confidence: 0.34615384615384615
Lift: 3.218202589489718
----------------------------------------------
Rule: frozenset({'38'}) -> frozenset({'12.0'})
Support: 0.008519701810436636
Confidence: 0.33333333333333337
Lift: 3.068627450980393
----------------------------------------------
Rule: frozenset({'44'}) -> frozenset({'11.0'})
Support: 0.005324813631522897
Confidence: 0.29411764705882354
Lift: 2.907120743034056
----------------------------------------------
Rule: frozenset({'19'}) -> frozenset({'30.0'})
Support: 0.010649627263045794
Confidence: 0.3448275862068966
Lift: 2.8910098522167487
----------------------------------------------
Rule: frozenset({'21'}) -> frozenset({'41.0'})
Support: 0.009584664536741214
Confidence: 0.34615384615384615
Lift: 2.87644656228727
----------------------------------------------
Rule: frozenset({'23'}) -> frozenset({'45.0'})
Support: 0.008519701810436636
Confidence: 0.4
Lift: 2.8454545454545457
----------------------------------------------
Rule: frozenset({'5.0'}) -> frozenset({'4'})
Support: 0.01703940362087327
Confidence: 0.3404255319148936
Lift: 2.7796484736355227
----------------------------------------------

# problem2






Most frequently chosen, 즉 support 값이 가장 큰 itemset을 찾았습니다.


In [205]:
association_rules3 = apriori(transactions, min_support=0.027, max_length=2)
association_results3 = list(association_rules3)

association_results3 = list(filter(lambda x: len(x.items) > 1, association_results3))

In [206]:
print(len(association_results3))

1


In [207]:
print(association_results3)

[RelationRecord(items=frozenset({'33.0', '40.0'}), support=0.027689030883919063, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'33.0', '40.0'}), confidence=0.027689030883919063, lift=1.0), OrderedStatistic(items_base=frozenset({'33.0'}), items_add=frozenset({'40.0'}), confidence=0.19999999999999998, lift=1.41203007518797), OrderedStatistic(items_base=frozenset({'40.0'}), items_add=frozenset({'33.0'}), confidence=0.1954887218045113, lift=1.41203007518797)])]


In [208]:
for item in association_results3:

    print("Itemsets: " + str(item[0]))
    print("Support: " + str(item[1]))

Itemsets: frozenset({'33.0', '40.0'})
Support: 0.027689030883919063


min_support값을 여러값으로 조정하여 알고리즘을 실행해본 결과, 0.027을 넘는 두개의 item을 가진 itemset은 하나 뿐이며

## **Itemset = {33, 40}** 
이 그에 해당합니다.



---



---



# problem 3

2번 문제와 같이 가장 함께 선택된 빈도가 많은 숫자 집합일수록 support값이 높습니다.

따라서 apriori 알고리즘을 통해 얻어진 3개의 숫자를 가진 itemset들 중 support 값이 가장 큰 값을 찾아보았습니다.

In [198]:
association_rules2 = apriori(transactions, min_support=0.008, max_length=3) #  setting min_support, max item lengthin itemset
association_results2 = list(association_rules2)

association_results2 = list(filter(lambda x: len(x.items) > 2, association_results2)) # item num in itemset is bigger than 2

In [199]:
print(len(association_results2))

1


In [200]:
print(association_results2)

[RelationRecord(items=frozenset({'6', '18.0', '31.0'}), support=0.008519701810436636, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'6', '18.0', '31.0'}), confidence=0.008519701810436636, lift=1.0), OrderedStatistic(items_base=frozenset({'18.0'}), items_add=frozenset({'6', '31.0'}), confidence=0.06201550387596899, lift=3.2351421188630494), OrderedStatistic(items_base=frozenset({'31.0'}), items_add=frozenset({'6', '18.0'}), confidence=0.06299212598425198, lift=2.9574803149606304), OrderedStatistic(items_base=frozenset({'6'}), items_add=frozenset({'18.0', '31.0'}), confidence=0.08247422680412371, lift=4.0759631036353765), OrderedStatistic(items_base=frozenset({'18.0', '31.0'}), items_add=frozenset({'6'}), confidence=0.4210526315789474, lift=4.075963103635377), OrderedStatistic(items_base=frozenset({'6', '18.0'}), items_add=frozenset({'31.0'}), confidence=0.4, lift=2.9574803149606304), OrderedStatistic(items_base=frozenset({'6', '31.0'}), items_add=froz

In [202]:
for item in association_results2:

    print("Itemsets: " + str(item[0]))
    print("Support: " + str(item[1]))

Itemsets: frozenset({'6', '18.0', '31.0'})
Support: 0.008519701810436636


min_support값을 여러값으로 조정하여 알고리즘을 실행해본 결과, 0.008을 넘는 세개의 item을 가진 itemset은 하나 뿐이며

## **Itemset = {6, 18, 31}** 
이 그에 해당합니다.